frontend/components/common/InstantSearchProvider.tsx

```diff
@@ -110,3 +115,31 @@ function useCountRenders() {
    countRef.current++;
    return countRef.current;
 }
+
+type RefreshSearchResultsProps = {
+   apiKey: ApiKey;
+};
+
+/**
+ * Returns a component that takes in an API key and refreshes search results
+ * if the API key changes (but only refreshes once for each api key change.
+ */
+function useSearchRefresher() {
+   // Create the ref in the parent component so it can live across re-renders
+   const previousApiKeyRef = React.useRef<ApiKey|null>(null);
+
+   return function RefreshSearchResults({
+      apiKey,
+   }: RefreshSearchResultsProps): null {
+      const { refresh } = useSearchCache();
+      useSafeLayoutEffect(() => {
+         const isFirstRender = !previousApiKeyRef.current;
+         const hasApiKeyChanged = previousApiKeyRef.current !== apiKey;
+         if (hasApiKeyChanged && !isFirstRender) {
+            refresh();
+         }
+         previousApiKeyRef.current = apiKey;
+      }, [apiKey, refresh]);
```

Comment:
Checking my understanding here:
If this runs client side, this will be a `useLayoutEffect` which runs synchronously before paint. Is this what is preventing a flash of a blank screen as referenced in fb70b6ee686b601af8af9fc74e4fbceecb55eafd?
If this runs server side, this will be a `useEffect`, but we don't necessarily need to `await refresh()` or anything?

Reply:
> If this runs client side, this will be a useLayoutEffect which runs synchronously before paint. Is this what is preventing a flash of a blank screen as referenced in https://github.com/iFixit/react-commerce/commit/fb70b6ee686b601af8af9fc74e4fbceecb55eafd?

I'll experiment with `useEffect()` instead, but this was mostly to decrease the time before we show the correct results, i.e. clear the cache sooner.

> If this runs server side, this will be a useEffect, 

useEffect() calls don't run their callbacks on the server as I understand it: https://codewithhugo.com/react-useeffect-ssr/

Reply:
> useEffect() calls don't run their callbacks on the server as I understand it

Oh I didn't know that. I think I misunderstand what `useSafeLayoutEffect` does then.

> this was mostly to decrease the time before we show the correct results, i.e. clear the cache sooner.

Cool, thanks for the elaboration

Reply:
https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-safe-layout-effect.ts